File: apps/android/app/src/main/java/chat/rogi/rogichat/channelport/core/model/channel/Channel.kt
package chat.rogi.rogichat.channelport.core.model.channel

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

// MARK: - Channel Verification Summary (per-platform)
data class ChannelVerificationSummary(
    val platform: String,
)

// MARK: - Domain Model
data class Channel(
    val id: Int,
    val name: String,
    val webPath: String,
    val platformUrl: String? = null,
    val profileImageUrl: String? = null,
    val topBannerUrl: String? = null,
    val themeColor: String = "#6366f1",
    val channelDescription: String? = null,
    val additionalLinks: List<ChannelLink> = emptyList(),
    val songCount: Int = 0,
    val artistCount: Int = 0,
    val categoryCount: Int = 0,
    val favoritesCount: Int = 0,
    val createdAt: String? = null,
    val updatedAt: String? = null,
    val isOwnerProSubscriber: Boolean = false,
    val verifications: List<ChannelVerificationSummary>? = null,
    val voiceCommissionActive: Boolean = false,
)

@Serializable
data class ChannelLink(
    val id: Int? = null,
    val name: String,
    val url: String,
    val icon: String? = null,
)

@Serializable
data class ChannelWardrobeResponse(
    val categories: List<ChannelWardrobeCategory> = emptyList(),
    val items: List<ChannelWardrobeItem> = emptyList(),
)

@Serializable
data class ChannelWardrobeCategory(
    val id: Int,
    val name: String,
    val defaultAspectRatio: String = "1:1",
    val isEnabled: Boolean = true,
    val order: Int = 0,
)

@Serializable
data class ChannelWardrobeItem(
    val id: Int,
    val categoryId: Int,
    val title: String,
    val imageUrl: String,
    val description: String? = null,
    val tags: List<String> = emptyList(),
    val isVisible: Boolean = true,
    val order: Int = 0,
    val createdAt: String? = null,
    val updatedAt: String? = null,
)

// MARK: - Channel Verification DTO (API Response)
@Serializable
data class ChannelVerificationDTO(
    val platform: String,
)

// MARK: - Channel DTO (API Response)
@Serializable
data class ChannelDTO(
    val id: Int,
    val name: String,
    val webPath: String,
    val platformUrl: String? = null,
    val profileImageUrl: String? = null,
    val topBannerUrl: String? = null,
    val leftBannerUrl: String? = null,
    val rightBannerUrl: String? = null,
    val additionalLinks: List<ChannelLink>? = null,
    val themeColor: String? = null,
    val channelDescription: String? = null,
    val createdAt: String? = null,
    val updatedAt: String? = null,
    @SerialName("_count")
    val count: ChannelCountDTO? = null,
    // Direct fields from popular channels API
    val favoritesCount: Int? = null,
    val songsCount: Int? = null,
    val isOwnerProSubscriber: Boolean? = null,
    val verifications: List<ChannelVerificationDTO>? = null,
    val voiceCommissionActive: Boolean? = null,
) {
    fun toDomain(): Channel = Channel(
        id = id,
        name = name,
        webPath = webPath,
        platformUrl = platformUrl,
        profileImageUrl = profileImageUrl,
        topBannerUrl = topBannerUrl,
        themeColor = themeColor ?: "#6366f1",
        channelDescription = channelDescription,
        additionalLinks = additionalLinks ?: emptyList(),
        songCount = songsCount ?: count?.songs ?: 0,
        artistCount = count?.artists ?: 0,
        categoryCount = count?.categories ?: 0,
        favoritesCount = this.favoritesCount ?: count?.userFavorites ?: 0,
        createdAt = createdAt,
        updatedAt = updatedAt,
        isOwnerProSubscriber = isOwnerProSubscriber ?: false,
        verifications = verifications?.map { ChannelVerificationSummary(platform = it.platform) },
        voiceCommissionActive = voiceCommissionActive ?: false,
    )
}

@Serializable
data class ChannelCountDTO(
    val songs: Int? = null,
    val artists: Int? = null,
    val categories: Int? = null,
    val userFavorites: Int? = null,
)

// MARK: - Favorite Channel DTO (API Response for /v1/favorites/channels)
@Serializable
data class FavoriteChannelDTO(
    val id: Int,
    val channelId: Int,
    val channelName: String,
    val profileImageUrl: String? = null,
    val webPath: String,
    val themeColor: String = "#6366f1",
    val ownerNickname: String? = null,
    val createdAt: String? = null,
    val songCount: Int? = null,
    val artistCount: Int? = null,
    val favoritesCount: Int? = null,
    val isOwnerProSubscriber: Boolean? = null,
) {
    fun toDomain(): Channel = Channel(
        id = channelId,
        name = channelName,
        webPath = webPath,
        profileImageUrl = profileImageUrl,
        themeColor = themeColor,
        songCount = songCount ?: 0,
        artistCount = artistCount ?: 0,
        favoritesCount = favoritesCount ?: 0,
        createdAt = createdAt,
        isOwnerProSubscriber = isOwnerProSubscriber ?: false,
    )
}

// MARK: - My Channel (채널 소유/관리 정보)
data class MyChannel(
    val id: Int,
    val name: String,
    val webPath: String,
    val profileImageUrl: String? = null,
    val themeColor: String = "#6366f1",
    val isOwner: Boolean,
)

@Serializable
data class MyChannelDTO(
    val id: Int,
    val name: String,
    val webPath: String,
    val platformUrl: String? = null,
    val profileImageUrl: String? = null,
    val topBannerUrl: String? = null,
    val leftBannerUrl: String? = null,
    val rightBannerUrl: String? = null,
    val additionalLinks: List<ChannelLink>? = null,
    val themeColor: String? = null,
    val channelDescription: String? = null,
    val createdAt: String? = null,
    val updatedAt: String? = null,
    @SerialName("_count")
    val count: ChannelCountDTO? = null,
    val isOwner: Boolean = false,
) {
    fun toDomain(): MyChannel = MyChannel(
        id = id,
        name = name,
        webPath = webPath,
        profileImageUrl = profileImageUrl,
        themeColor = themeColor ?: "#6366f1",
        isOwner = isOwner,
    )
}

// MARK: - Channel Permission
@Serializable
data class ChannelPermission(
    val view: Boolean = true,
    val manageContent: Boolean = false,
    val manageSettings: Boolean = false,
    val isOwner: Boolean = false,
)

// MARK: - Channel Profile DTO (API Response)
@Serializable
data class ChannelProfile(
    val channelId: Int? = null,
    val nickname: String? = null,
    val birthday: String? = null,
    val residence: String? = null,
    val nationality: String? = null,
    val gender: String? = null,
    val heightCm: String? = null,
    val weightKg: String? = null,
    val mbti: String? = null,
    val symbolColor: String? = null,
    val agency: String? = null,
    val affiliatedGroups: List<String>? = null,
    val fandomName: String? = null,
    val religion: String? = null,
    val description: String? = null,
    val homeDescription: String? = null,
    val education: List<String>? = null,
    val alias: List<String>? = null,
    val debutDate: String? = null,
    val broadcastingPlatforms: List<String>? = null,
    val bio: String? = null,
    val links: List<ProfileLink>? = null,
    val updatedAt: String? = null,
    val anniversaries: ProfileAnniversaries? = null,
) {
    // Computed property for display
    val height: String? get() = heightCm
}

@Serializable
data class ProfileLink(
    val label: String? = null,
    val url: String,
    val icon: String? = null,
)

@Serializable
data class ProfileAnniversaries(
    val milestones: ProfileMilestones? = null,
    val birthday: ProfileBirthday? = null,
    val nextUpcomingEvent: ProfileUpcomingEvent? = null,
)

@Serializable
data class ProfileMilestones(
    val daysPassed: Int = 0,
    val nextMilestone: String? = null,
    val daysToMilestone: Int = 0,
)

@Serializable
data class ProfileBirthday(
    val daysUntilBirthday: Int = 0,
    val birthdayDate: String? = null,
)

@Serializable
data class ProfileUpcomingEvent(
    val type: String? = null,
    val label: String? = null,
    val daysUntil: Int = 0,
)

// MARK: - Update Channel Request
@Serializable
data class UpdateChannelRequest(
    val name: String,
    val webPath: String,
    val profileImageUrl: String? = null,
    val additionalLinks: List<ChannelLink> = emptyList(),
    val themeColor: String,
    val channelDescription: String? = null,
)

File: apps/ios/Sources/Features/Channel/Channel.swift
import Foundation

struct Channel: Identifiable, Equatable {
    let id: Int
    let name: String
    let webPath: String
    let platformUrl: String?
    let profileImageUrl: String?
    let themeColor: String
    let channelDescription: String?
    let additionalLinks: [ChannelLink]
    let songCount: Int
    let artistCount: Int
    let categoryCount: Int
    let favoritesCount: Int
    let verifications: [ChannelVerificationSummary]
    let isOwnerProSubscriber: Bool
    let voiceCommissionActive: Bool
    let createdAt: Date
    let updatedAt: Date
}

struct ChannelLink: Codable, Equatable {
    let name: String
    let url: String
}

struct ChannelVerificationSummary: Codable, Equatable, Identifiable {
    let platform: String

    var id: String { platform }
    var isApproved: Bool { true } // API only returns approved verifications
}

struct ChannelSummary: Identifiable, Equatable {
    let id: Int
    let name: String
    let webPath: String
    let profileImageUrl: String?
    let themeColor: String
}

// MARK: - My Channel (채널 소유/관리 정보)
struct MyChannel: Identifiable, Equatable {
    let id: Int
    let name: String
    let webPath: String
    let profileImageUrl: String?
    let themeColor: String
    let isOwner: Bool
}

struct MyChannelDTO: Decodable {
    let id: Int
    let name: String
    let webPath: String
    let platformUrl: String?
    let profileImageUrl: String?
    let topBannerUrl: String?
    let leftBannerUrl: String?
    let rightBannerUrl: String?
    let additionalLinks: [ChannelLink]?
    let themeColor: String?
    let channelDescription: String?
    let createdAt: String?
    let updatedAt: String?
    let _count: ChannelCountDTO?
    let isOwner: Bool

    struct ChannelCountDTO: Decodable {
        let songs: Int?
        let artists: Int?
        let categories: Int?
    }

    func toDomain() -> MyChannel {
        MyChannel(
            id: id,
            name: name,
            webPath: webPath,
            profileImageUrl: profileImageUrl,
            themeColor: themeColor ?? "#6366f1",
            isOwner: isOwner
        )
    }
}

// MARK: - Channel DTO
struct ChannelDTO: Decodable {
    let visibility: String?
    let id: Int
    let name: String
    let webPath: String
    let platformUrl: String?
    let profileImageUrl: String?
    let topBannerUrl: String?
    let leftBannerUrl: String?
    let rightBannerUrl: String?
    let additionalLinks: [ChannelLink]?
    let themeColor: String?
    let channelDescription: String?
    let verifications: [ChannelVerificationSummary]?
    let isOwnerProSubscriber: Bool?
    let voiceCommissionActive: Bool?
    let createdAt: String?
    let updatedAt: String?
    let _count: ChannelCountDTO?
    /// 최신 인기 목록 API는 집계된 즐겨찾기 수를 `_count` 밖에 내려줍니다.
    let favoritesCount: Int?
    let popularityScore: Double?

    struct ChannelCountDTO: Decodable {
        let songs: Int?
        let artists: Int?
        let categories: Int?
        let userFavorites: Int?
    }

    /// `favoriteCountFromPopularityScore`는 즐겨찾기 가중치만 1로 요청한
    /// 채널 목록에서만 사용합니다. 그 요청의 popularityScore는 즐겨찾기 수와 같습니다.
    func toDomain(favoriteCountFromPopularityScore: Bool = false) -> Channel {
        let dateFormatter = ISO8601DateFormatter()
        dateFormatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        let effectiveFavoritesCount: Int
        if let favoritesCount {
            effectiveFavoritesCount = favoritesCount
        } else if favoriteCountFromPopularityScore, let popularityScore {
            effectiveFavoritesCount = Int(popularityScore.rounded())
        } else {
            effectiveFavoritesCount = _count?.userFavorites ?? 0
        }

        return Channel(
            id: id,
            name: name,
            webPath: webPath,
            platformUrl: platformUrl,
            profileImageUrl: profileImageUrl,
            themeColor: themeColor ?? "#6366f1",
            channelDescription: channelDescription,
            additionalLinks: additionalLinks ?? [],
            songCount: _count?.songs ?? 0,
            artistCount: _count?.artists ?? 0,
            categoryCount: _count?.categories ?? 0,
            favoritesCount: effectiveFavoritesCount,
            verifications: verifications ?? [],
            isOwnerProSubscriber: isOwnerProSubscriber ?? false,
            voiceCommissionActive: voiceCommissionActive ?? false,
            createdAt: createdAt.flatMap { dateFormatter.date(from: $0) } ?? Date(),
            updatedAt: updatedAt.flatMap { dateFormatter.date(from: $0) } ?? Date()
        )
    }

}

// MARK: - Favorite Channel Response
struct FavoriteChannelDTO: Decodable {
    let id: Int
    let channelId: Int
    let channelName: String
    let profileImageUrl: String?
    let webPath: String
    let themeColor: String
    let ownerNickname: String
    let channelDescription: String?
    let createdAt: String
    let songCount: Int?
    let artistCount: Int?
    let favoritesCount: Int?

    func toChannel() -> Channel {
        let dateFormatter = ISO8601DateFormatter()
        dateFormatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        let createdDate = dateFormatter.date(from: createdAt) ?? Date()

        return Channel(
            id: channelId,
            name: channelName,
            webPath: webPath,
            platformUrl: nil,
            profileImageUrl: profileImageUrl,
            themeColor: themeColor,
            channelDescription: nil,
            additionalLinks: [],
            songCount: songCount ?? 0,
            artistCount: artistCount ?? 0,
            categoryCount: 0,
            favoritesCount: favoritesCount ?? 0,
            verifications: [],
            isOwnerProSubscriber: false,
            voiceCommissionActive: false,
            createdAt: createdDate,
            updatedAt: createdDate
        )
    }
}

struct FavoriteChannelsResponse: Decodable {
    let items: [FavoriteChannelDTO]
    let total: Int
    let page: Int
    let limit: Int
    let totalPages: Int

    enum CodingKeys: String, CodingKey {
        case items = "favorites"
        case total
        case page
        case limit
        case totalPages
    }
}

// MARK: - Channel Profile
struct ChannelProfile: Decodable {
    let channelId: Int
    let birthday: String?
    let residence: String?
    let heightCm: String?
    let weightKg: String?
    let nationality: String?
    let gender: String?
    let symbolColor: String?
    let agency: String?
    let affiliatedGroups: [String]?
    let fandomName: String?
    let religion: String?
    let nickname: String?
    let description: String?
    let homeDescription: String?
    let education: [String]?
    let mbti: String?
    let alias: [String]?
    let debutDate: String?
    let broadcastingPlatforms: [String]?
    let bio: String?
    let links: [ProfileLink]?
    let updatedAt: String?
    let anniversaries: ProfileAnniversaries?

    struct ProfileLink: Decodable {
        let label: String?
        let url: String
        let icon: String?
    }

    struct ProfileAnniversaries: Decodable {
        let milestones: ProfileMilestones?
        let birthday: ProfileBirthday?
        let nextUpcomingEvent: ProfileUpcomingEvent?
    }

    struct ProfileMilestones: Decodable {
        let daysPassed: Int
        let nextMilestone: String
        let daysToMilestone: Int
    }

    struct ProfileBirthday: Decodable {
        let daysUntilBirthday: Int
        let birthdayDate: String
    }

    struct ProfileUpcomingEvent: Decodable {
        let type: String
        let label: String
        let daysUntil: Int
    }

    // 성별을 한글로 변환
    var genderDisplayName: String? {
        guard let gender = gender else { return nil }
        switch gender {
        case "MALE": return "남성"
        case "FEMALE": return "여성"
        case "NONBINARY": return "논바이너리"
        case "OTHER": return "기타"
        case "SECRET": return "비공개"
        default: return gender
        }
    }

    // 생일을 표시 형식으로 변환
    var birthdayDisplayName: String? {
        guard let birthday = birthday else { return nil }
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withFullDate]
        guard let date = formatter.date(from: birthday) else { return birthday }

        let displayFormatter = DateFormatter()
        displayFormatter.locale = Locale(identifier: "ko_KR")

        let calendar = Calendar.current
        let year = calendar.component(.year, from: date)

        if year == 1900 {
            displayFormatter.dateFormat = "M월 d일"
        } else {
            displayFormatter.dateFormat = "yyyy년 M월 d일"
        }

        return displayFormatter.string(from: date)
    }

    // 데뷔일을 표시 형식으로 변환
    var debutDateDisplayName: String? {
        guard let debutDate = debutDate else { return nil }
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        guard let date = formatter.date(from: debutDate) else { return debutDate }

        let displayFormatter = DateFormatter()
        displayFormatter.locale = Locale(identifier: "ko_KR")
        displayFormatter.dateFormat = "yyyy년 M월 d일"

        return displayFormatter.string(from: date)
    }
}
